mitodl/ocw-studio
# update CKEditor config to use our resource-link fork

#### Pre-Flight checklist



#### What are the relevant tickets?

closes #607
closes #612
part of #536

#### What's this PR do?

This PR changes how we support linking to resources. Instead of a block-level element (which is sort of like a paragraph in terms of how it fits in to the document) we now support inline linking. Basically, this means that the resource link is actually an attribute on a span of text, instead of a block element, and so the text inside of it can be edited and so on.

The user experience is not _the greatest_ at present, but here's how it works:

- the user highlights some text in the editor
- the user clicks 'Link resource'
- the user uses the resource picker dialog to pick a resource, and clicks the 'Link resource' confirmation button
- a resource link is added to the editor state, serializing to the `{{< resource_link >}}` shortcode

To get this all working I had to make a few other changes. Some things of note:

- we now have a fork of ckeditor here: https://github.com/mitodl/ckeditor. this is being used just to fork the `Link` plugin, but since ckeditor is developed as a monorepo we need to fork the whole thing. The changes I've made to the plugin to support our weird use-case are here: https://github.com/mitodl/ckeditor5/pull/1 Once this PR is approved I will merge that PR as well and publish a non-beta version of our custom package to npm.
- I changed how the resource picker works a little bit. Instead of having one 'Add resource' button we now have 'Embed resource' and 'Link resource' buttons. Together these make it a little less ambiguous for the user (at least, that is the hope!)

#### How should this be manually tested?

follow the steps I outlined above. Keep in mind that the UX is still basic now. We do not yet support editing which resource a resource link points to once it is created, for instance. For now the user should just delete and recreate to do that. But generally play around with it and make sure it at least does what it's supposed to.

#### Screenshots (if appropriate)

<img width="730" alt="Screen Shot 2021-09-28 at 12 38 23 PM" src="https://user-images.githubusercontent.com/6207644/135129240-92bf753a-efa3-4f41-9d64-c6c91133c0dc.png">
